File: Algorithms/0929_Unique_Email_Addresses/Python/Unique_Email_Addresses_Solution_1.py

# Space: O(n)
# Time: O(n)

class Solution:
    def numUniqueEmails(self, emails):
        length = len(emails)
        if length == 1: return 1
        temp = set()
        for each in emails:
            temp.add(self.get_real_name(each))
        return len(temp)

    def get_real_name(self, string: str):
        temp = string.split('@')
        temp[0] = temp[0].replace('.', '')
        if '+' in temp[0]:
            temp[0] = temp[0][:temp[0].index('+')]
        res = temp[0] + '@' + temp[1]
        return res

File: Algorithms/0929_Unique_Email_Addresses/Java/UniqueEmailAddressesSolution1.java

// Space: O(n)
// Time: O(n)

class Solution {
    public int numUniqueEmails(String[] emails) {
        Set<String> res = new HashSet<>();
        for (String email : emails) {
            res.add(cleanEmail(email));
        }
        return res.size();
    }

    private String cleanEmail(String email) {
        StringBuilder res = new StringBuilder();
        int index = email.indexOf('@');
        for (char c : email.toCharArray()) {
            if (c == '@' || c == '+') break;
            else if (c != '.') res.append(c);
        }
        res.append(email.substring(index));
        return res.toString();
    }
}




